Initial Multichain API docs (#1621)
MetaMask/metamask-docs: wallet/reference/multichain-api.md

:::note See also
- [About the Multichain API](../concepts/multichain-api.md)
- [Interact with multiple networks](../how-to/manage-networks/use-multichain.md)
:::

## Methods

### `wallet_createSession`

Creates a multichain connection with a wallet, authorizing that wallet with the specified set of scopes
and properties.
This method is defined in [CAIP-25](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md).

#### Parameters

- `optionalScopes`: `object` - (Optional) [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) authorization scopes the wallet can support in order to be used with this dapp.
  If scopes are specified, only the following properties are supported:
  - `references`: `array` - (Optional) A list of references to specific blockchains for the namespace of this scope.
    This property can only be used if the scope namespace does not already specify the blockchain.
    For example, you can use this property for an `"eip155"` scope, but not an `"eip155:10"` scope.
    
    References are mainly used when there would otherwise be duplicate scopes.

>   - `methods`: `array` - A list of JSON-RPC methods the wallet must support in order to be used with this dapp.

Reviewer: ```suggestion
  - `methods`: `array` - A list of JSON-RPC methods the wallet must support to be compatible with the dapp.
```

>   - `notifications`: `array` - A list of JSON-RPC notifications the wallet must support in order to be used with this dapp.

Reviewer: ```suggestion
  - `notifications`: `array` - A list of JSON-RPC notifications the wallet must support to be compatible with the dapp.
```

  - `accounts`: `array` - (Optional) A list of [CAIP-10](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-10.md)
    account IDs valid within this scope.

>     Dapps should only supply this property when they know the set of accounts they would like the user to permit.

Reviewer: ```suggestion
    Dapps should include this only if they know which accounts they want the user to permit.
```

    When supplied, these accounts are preselected by default in the account selection process.
    Dapps typically omit this property for the user to select their own accounts.
- `sessionProperties`: `object` - (Optional) Properties that the wallet can use to determine if the connection is valid.

> - `requiredScopes`: `object` - (Optional) [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) authorization scopes the wallet must support in order to be used
>   with this dapp.

Reviewer: ```suggestion
- `requiredScopes`: `object` - (Optional) [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) authorization scopes the wallet must support to be compatible with the dapp.
```

  - `accounts`: `array` - (Optional) A list of [CAIP-10](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-10.md)
    account IDs valid within this scope.
    Dapps should only supply this property when they know the set of accounts they would like the user to permit.
    When supplied, these accounts are preselected by default in the account selection process.
    Dapps typically omit this property for the user to select their own accounts.
- `sessionProperties`: `object` - (Optional) Properties that the wallet can use to determine if the connection is valid.
- `requiredScopes`: `object` - (Optional) [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) authorization scopes the wallet must support in order to be used
  with this dapp.
  We don't recommend using `requiredScopes` with MetaMask.
  MetaMask treats all `requiredScopes` as `optionalScopes`.

#### Returns

The scopes and properties of the created connection.

#### Example

<Tabs>
<TabItem value="Request">

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "wallet_createSession",
  "params": {
    "optionalScopes": {
      "eip155": {
        "references": ["1", "137"],

>         "methods": ["eth_sendTransaction", "eth_signTransaction", "eth_sign", "get_balance", "personal_sign"],

Reviewer 2: ```suggestion
        "methods": ["eth_sendTransaction", "eth_signTransaction", "eth_sign", "eth_subscribe", "get_balance", "personal_sign"],
```

maybe? It's weird to have `eth_subscription` without a way to use it

        "notifications": ["eth_subscription"]
      },
      "eip155:10": {
        "methods": ["get_balance"],
        "notifications": []
      },
      "eip155:0": {
        "methods": ["wallet_getPermissions", "wallet_creds_store", "wallet_creds_verify", "wallet_creds_issue", "wallet_creds_present"],
        "notifications": []
      },
      "eip155:42161": {
        "methods": ["eth_sendTransaction", "eth_signTransaction", "get_balance", "personal_sign"],
        "notifications": []
      }
    }
  }
}
```

</TabItem>
<TabItem value="Result">

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "sessionScopes": {
      "eip155": {
        "references": ["1", "137"],

>         "methods": ["eth_sendTransaction", "eth_signTransaction", "get_balance", "eth_sign", "personal_sign"],

Reviewer 2: ```suggestion
        "methods": ["eth_sendTransaction", "eth_signTransaction", "eth_subscribe", "get_balance", "eth_sign", "personal_sign"],
```

        "notifications": ["eth_subscription"],
        "accounts": ["eip155:1:0xab16a96d359ec26a11e2c2b3d8f8b8942d5bfcdb", "eip155:137:0xab16a96d359ec26a11e2c2b3d8f8b8942d5bfcdb"]
      },
      "eip155:10": {
        "methods": ["get_balance"],
        "notifications": [],
        "accounts": []
      },
      "eip155:42161": {
        "methods": ["personal_sign"],
        "notifications": [],
        "accounts": ["eip155:42161:0x0910e12C68d02B561a34569E1367c9AAb42bd810"]
      },

>       "eip155:0": {

Reviewer 2: we don't support the null chain scope or the methods in this scope object. I think we should just remove it? 

That begs the question of if we should advertise that `wallet_addEthereumChain` exists in `wallet:eip155`. I'd argue no, not at the moment until we iron out our `wallet:<ecosystem>` scope usage

@adonesky1 @vandan

Author: Removed `eip155:0` scope object from example request and response.

Reviewer 3: @jiexi For Ethereum-oriented usage of the Multichain API, we will need to illustrate precisely how `wallet_addEthereumChain` can be used. That said, it's not critical for the initial version of these docs.

Reviewer 4: Yeah I would agree w/ @jiexi lets remove the `wallet_addEthereumChain` reference for now, since it seems like we may want to migrate away from `wallet:eip155` syntax

---
description: See the Multichain API reference.
sidebar_custom_props:
  flask_only: true
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

# Multichain API

:::flaskOnly
:::

Dapps can call Multichain API [methods](#methods) to create and manage
[CAIP-25](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md) connections with MetaMask.
The API also provides [events](#events) that wallets can send to dapps to notify them of onchain or connection changes.

:::note See also
- [About the Multichain API](../concepts/multichain-api.md)
- [Interact with multiple networks](../how-to/manage-networks/use-multichain.md)
:::

## Methods

### `wallet_createSession`

Creates a multichain connection with a wallet, authorizing that wallet with the specified set of scopes

> and properties.

Reviewer 2: scopes, accounts, methods, notifications, and properties? not sure if too specific or not

Author: If accounts, methods, and notifications are *part of* a scope, then I think the shorter description suffices. The parameters section goes into detail about what's included.

This method is defined in [CAIP-25](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md).

#### Parameters

- `optionalScopes`: `object` - (Optional) [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) authorization scopes the wallet can support in order to be used with this dapp.
  If scopes are specified, only the following properties are supported:

>   - `references`: `array` - (Optional) A list of references to specific blockchains for the namespace of this scope.

Reviewer 2: this can only be specified if the key of `optionalScopes` is just an ecosystem like `eip155`, so the following is valid
```
{
  'eip155': {
     references:['1', '5']  
   }
}
```

but this is not
```
{
  'eip155:1': {
     references:['1', '5']  
   }
}
```

Author: Added more description to clarify this. The example also shows the differences.

    This property is mainly used when there would otherwise be duplicate scopes.
  - `methods`: `array` - A list of JSON-RPC methods the wallet must support in order to be used with this dapp.
  - `notifications`: `array` - A list of JSON-RPC notifications the wallet must support in order to be used with this dapp.
  - `accounts`: `array` - (Optional) A list of [CAIP-10](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-10.md)
    account IDs valid within this scope.
    Dapps should only supply this property when they know the set of accounts they would like the user to permit.
    When supplied, these accounts are preselected by default in the account selection process.
    Dapps typically omit this property for the user to select their own accounts.
- `sessionProperties`: `object` - (Optional) Properties that the wallet can use to determine if the connection is valid.
- `requiredScopes`: `object` - (Optional) [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) authorization scopes the wallet must support in order to be used
  with this dapp.

>   We don't recommend using `requiredScopes` with MetaMask.

Reviewer 2: We treat all requiredScopes as optional despite the name

Author: Added to description.

#### Returns

The scopes and properties of the created connection.

#### Example

<Tabs>
<TabItem value="Request">

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "wallet_createSession",
  "params": {
    "optionalScopes": {
      "eip155": {
        "references": ["1", "137"],
        "methods": ["eth_sendTransaction", "eth_signTransaction", "eth_sign", "get_balance", "personal_sign"],
        "notifications": ["accountsChanged", "chainChanged"]
      },
      "eip155:10": {
        "methods": ["get_balance"],
        "notifications": ["accountsChanged", "chainChanged"]
      },
      "eip155:0": {
        "methods": ["wallet_getPermissions", "wallet_creds_store", "wallet_creds_verify", "wallet_creds_issue", "wallet_creds_present"],
        "notifications": []
      },
      "eip155:42161": {
        "methods": ["eth_sendTransaction", "eth_signTransaction", "get_balance", "personal_sign"],

>         "notifications": ["accountsChanged", "chainChanged"]

Reviewer 2: there are no `accountsChanged` or `chainChanged` notifications for EVM in the Multichain API

Reviewer 2: same thing for the result.

`eth_subscription` is the only valid notification for evm currently

Author: updated examples

      }
    },
    "sessionProperties": {
      "expiry": "2022-12-24T17:07:31+00:00",

>       "caip154-mandatory": "true"

Reviewer 2: @adonesky1 we technically don't enforce these yet

Reviewer 4: Yeah let's remove this

Reviewer 2: yeahh... hmm... let's remove `sessionProperties` entirely until we actually support it to avoid confusion. Thank you for clarifying!

#### Example

<Tabs>
<TabItem value="Request">

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "wallet_getSession",
  "params": {}
}
```

</TabItem>
<TabItem value="Result">

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "sessionScopes": {
      "eip155:1": {
        "methods": ["eth_signTransaction"],
        "notifications": ["accountsChanged"],
        "accounts": ["eip155:1:0xabc123"]
      },
      "eip155:137": {
        "methods": ["eth_sendTransaction"],
        "notifications": ["chainChanged"],
        "accounts": ["eip155:137:0xdef456"]
      },
      "solana:4sGjMW1sUnHzSxGspuhpqLDx6wiyjNtZ": {
        "methods": ["getBalance", "getAccountInfo", "sendTransaction", "getBlock"],
        "notifications": [],
        "accounts": ["solana:4sGjMW1sUnHzSxGspuhpqLDx6wiyjNtZ:4Nd1mS8AUwK3kU3gdiAM6QCvqhA7Do8rKtMXsGyqrJxy"]
      }
    }
  }
}
```

</TabItem>
</Tabs>

### `wallet_invokeMethod`

Invokes the specified [JSON-RPC API](json-rpc-methods/index.md) method on the specified network
previously authorized to the dapp within a connection.
This method is defined in [CAIP-27](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-27.md).

:::note

> MetaMask doesn't support session IDs.

Reviewer 2: do we need this note for getSession as well?

Author: The same note is already in the description of `wallet_getSession`.
